winml/lib/Api.Image/TensorToVideoFrameConverter.cpp

```diff
@@ -593,6 +596,54 @@ void TensorToVideoFrameConverter::ConvertGPUTensorToSoftwareBitmap(
   readback_heap_->Unmap(0, &CD3DX12_RANGE(0, 0));
 }
 
+void TensorToVideoFrameConverter::ConvertBatchedDX12TensorToBuffers(
+    ID3D12Resource* input_tensor,
+    size_t buffer_size_in_bytes,
+    _winml::D3DDeviceCache& device_cache,
+    const std::vector<wss::IBuffer>& buffers) {
+  assert(input_tensor != nullptr);
+
+  // TODO: Make an allocator for readback heaps
+  if (!readback_heap_ || readback_heap_->GetDesc().Width < buffer_size_in_bytes) {
+    WINML_THROW_IF_FAILED(device_cache.GetD3D12Device()->CreateCommittedResource(
+        &CD3DX12_HEAP_PROPERTIES(D3D12_HEAP_TYPE_READBACK),
+        D3D12_HEAP_FLAG_NONE,
+        &CD3DX12_RESOURCE_DESC::Buffer(buffer_size_in_bytes),
+        D3D12_RESOURCE_STATE_COPY_DEST,
+        nullptr,
+        IID_PPV_ARGS(&readback_heap_)));
+  }
+
+  ResetCommandList(device_cache);
+
+  auto barrier = CD3DX12_RESOURCE_BARRIER::Transition(input_tensor, D3D12_RESOURCE_STATE_UNORDERED_ACCESS, D3D12_RESOURCE_STATE_COPY_SOURCE);
+  command_list_->ResourceBarrier(1, &barrier);
+  command_list_->CopyBufferRegion(readback_heap_.Get(), 0, input_tensor, 0, buffer_size_in_bytes);
+
+  WINML_THROW_IF_FAILED(command_list_->Close());
+  ID3D12CommandList* ppCommandLists[] = {command_list_.Get()};
+  device_cache.GetCommandQueue()->ExecuteCommandLists(_countof(ppCommandLists), ppCommandLists);
+
+  // Sync to make sure the the heap received all the data
+  device_cache.SyncD3D12ToCPU();
+
+  byte* readback_buffer = nullptr;
+  WINML_THROW_IF_FAILED(readback_heap_->Map(0, &CD3DX12_RANGE(0, buffer_size_in_bytes), reinterpret_cast<void**>(&readback_buffer)));
+  auto readback_buffer_span = gsl::span<byte>(readback_buffer, buffer_size_in_bytes);
+  _winml::LoadOrStoreDisjointBuffers(
```

Comment:
Nit: Might be cleaner to make two different APIs for Load and Store, and handle the branching logic in the implementation.

Reply:
Can do in a follow up change